src/plugins/AvatarPEPPlugin.ts

```diff
@@ -0,0 +1,245 @@
+         avatarUI.reload();
+      }
+
+      return true;
+   };
+
+   private avatarProcessor = async (contact: IContact, avatar: IAvatar): Promise<[IContact, IAvatar]> => {
+      let storage = this.getStorage();
+      let hash = storage.getItem(contact.getJid().bare);
+
+      if (!hash && !avatar) {
+         try {
+            const avatarObject = await this.getAvatar(contact);
+            const data = avatarObject.src.replace(/^.+;base64,/, '');
+            avatar = new Avatar(Hash.SHA1FromBase64(data), avatarObject.type, avatarObject.src);
+
+            this.getStorage().setItem(contact.getJid().bare, avatar.getHash() || '');
+
+            let avatarUI = AvatarUI.get(contact);
+            avatarUI.reload();
+         } catch (err) {
+            // we could not find any avatar
+         }
+      }
+
+      if (!hash || avatar) {
+         return [contact, avatar];
+      }
+
+      try {
+         avatar = new Avatar(hash);
+      } catch (err) {
+         try {
+            const avatarObject = await this.getAvatar(contact);
+
+            if (avatarObject) {
+               avatar = new Avatar(hash, avatarObject.type, avatarObject.src);
+               return [contact, avatar];
+            } else {
+               this.pluginAPI.Log.warn('No local cached avatar found');
+               return [contact, avatar];
+            }
+         } catch (err) {
+            this.pluginAPI.Log.warn('Error during avatar retrieval', err);
+            return [contact, avatar];
+         }
+      }
+
+      return [contact, avatar];
+   };
+
+   private async getAvatar(contact: IContact): Promise<{ src: string; type: string }> {
+      let connection = this.pluginAPI.getConnection();
+
+      return connection
+         .getPEPService()
+         .retrieveItems(UANS_METADATA, contact.getJid().bare)
+         .then(meta => {
+            let metadata = $(meta).find('metadata[xmlns="urn:xmpp:avatar:metadata"]');
+
+            if (metadata.length > 0) {
+               let info = metadata.find('info');
+
+               if (info && info.length > 0) {
+                  let hash = $(info).attr('id');
+                  if (hash && hash.length > 0) {
+                     let typeval = $(info).attr('type');
+
+                     let result = connection
+                        .getPEPService()
+                        .retrieveItems(UANS_DATA, contact.getJid().bare)
+                        .then(data => {
+                           if (data && $(data).text() && $(data).text().trim().length > 0) {
+                              let src = $(data)
+                                 .text()
+                                 .replace(/[\t\r\n\f]/gi, '');
+                              const b64str = src.replace(/^.+;base64,/, '');
+                              this.getStorage().setItem(contact.getJid().bare, Hash.SHA1FromBase64(b64str));
+                              return { src: 'data:' + typeval + ';base64,' + src, type: typeval };
```

Comment:
We need to check if type and src is valid.

Reply:
ok we could check if sha of src equals to known sha hash but checking the type? detecting mimetype via magic numbers?

Reply:
I only want to prevent XSS, so we should check if typeval is a valid mimetype and src contains only valid base64 characters.

Reply:
i have added some checks now:
check data source for jpeg/png file
check for valid base64 characters
check mimetype for jpeg or png